_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+
+In this post, I'll try to explain the main concepts behind our programming model. Here they are, in chronological order:
+  - Bind-less draw calls
+  - Command buffers
+  - Resource views
+  - Pipeline states
+
+
+## Bind-less Draw Calls
+
+Low-level graphics operates with states, very much like low-level assembly mutates registers. In GL, you'd set a bunch of states and then issue a draw call like this:
+{% highlight Rust %}
+gl.UseProgram(program);
+gl.EnableVertexAttrib(0);
+gl.VertexAttribPointer(...);
+gl.Enable(gl::BLEND);
+gl.DrawArrays(gl::LINES, 0, 10);
+// restore states
+{% endhighlight %}
+
+This approach is painful due to the lack of locality. Seeing a draw call code would not allow you to predict the outcome easily, since you'd have to inspect all the states that are being used and where they are set up prior to the draw call. It's a classic imperative programming with all its faults.
+
+With GFX, you pass everything into the draw call:
+{% highlight Rust %}
+encoder.draw(
+	&slice, // contains the range of vertices and (optionally) an index buffer pointer
+	&pso,	// the pipeline state object, having all the graphics state encapsulated
+	&data	// the pipeline data, such as: vertex buffers, constant buffers, textures, etc
+);
+{% endhighlight %}
+
+Note that you don't pass *all* the state in each draw call, but only what is needed. We encode as much dependency information as possible into the type system, while still allowing one to trade the run-time correctness for the compile-time in case where more flexibility is needed.
+
+
+## Command Buffers
+
+Last-gen APIs (such as OpenGL and Dx11) suffered from the limited multi-threading capabilities, because you could only render in one thread that owned the graphics context. The CPU rendering overhead thus would often become the bottleneck of applications, reducing the number of draw calls sent through the pipeline.
+
+This issue has been properly addressed by all of the current-gen APIs (Dx12, Vulkan, etc) - they introduced the concept of a command buffer that can be filled up independently of the rendering context. Conceptually, a command buffer is an opaque object that is optimized for direct execution on your graphics hardware (it's not portable, much like the machine code). The idea is - you can populate multiple command buffers on multiple threads simultaneously, and then only "submit" them to the single graphics queue that would execute them on the hardware. Obviously, the concept complicates programming a bit. In exchange, it allows multi-threading your renderer and thus drawing many more objects on screen.
```

Comment:
> and then only "submit" them to the single graphics queue that would execute them on the hardware.

only? what do the quotation marks imply? maybe schedule instead?

> and thus drawing many more objects on screen.

+ being able to draw